sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/FindLatestForkWalk.java

```diff
@@ -0,0 +1,178 @@
+import org.eclipse.jgit.revwalk.RevWalk;
+
+/**
+ * Walk the graph starting from the HEAD of all references (branches).
+ * The idea is to find the most recent commit that can be reached from the current head and from another branch.
+ *
+ * This is a special walk that is not supported by RevWalk in JGit because we have the following requirements:
+ * 1) Might involve very high number of references (MergeBase detection in RebWalk can only involve limited number of branches)
+ * 2) We want any merge base involving the current branch and any other branch (MergeBase would try to detect a n-way merge base involving all branches)
+ * 3) We are not interested in the case where the head is merged into another branch (is a parent)
+ */
+public class FindLatestForkWalk {
+  private final DateRevQueue queue = new DateRevQueue();
+  private final Map<RevCommit, List<Branch>> commitOrigins = new HashMap<>();
+  private final Repository repo;
+
+  public FindLatestForkWalk(Repository repo) {
+    this.repo = repo;
+  }
+
+  @CheckForNull
+  public ForkPoint find() throws IOException {
+    try (RevWalk walk = new RevWalk(repo)) {
+      Branch currentBranch = new Branch(walk.parseCommit(repo.exactRef("HEAD").getObjectId()), "HEAD");
+      Set<Branch> refs = allRefsExceptCurrent(walk, currentBranch.head);
+      init(walk, currentBranch, refs);
+
+      RevCommit next;
+      while ((next = queue.next()) != null) {
+        List<Branch> origins = commitOrigins.get(next);
+
+        if (origins.size() > 1 && origins.contains(currentBranch)) {
+          return matchFound(next, origins);
+        }
+
+        for (RevCommit p : next.getParents()) {
+          if (p.equals(currentBranch.head)) {
+            // we are not interested in branches where the current branch is part of
+            continue;
+          }
+
+          // carry origins to parents, even if it has been seen before
+          registerOrigins(p, origins);
+
+          if (!p.has(RevFlag.SEEN)) {
+            // parse now to have commit time for the queue
+            parse(walk, p);
+            p.add(RevFlag.SEEN);
+            queue.add(p);
+          }
+        }
+
+        commitOrigins.remove(next);
+      }
+    }
+
+    return null;
+  }
+
+  private static ForkPoint matchFound(RevCommit commit, List<Branch> origins) {
+    Set<String> references = origins.stream()
+      .map(b -> b.name)
+      .filter(n -> !"HEAD".equals(n))
+      .map(n -> n.startsWith("refs/heads/") ? n.substring(11) : n)
+      .collect(Collectors.toSet());
+    return new ForkPoint(commit.getName(), references);
+  }
+
+  private static class Branch {
+    private final RevCommit head;
+    private final String name;
+
+    private Branch(RevCommit head, String name) {
+      this.head = head;
+      this.name = name;
+    }
+  }
+
+  private Optional<RevCommit> refToCommit(RevWalk walk, Ref ref) throws IOException {
+    RevObject maybehead = walk.parseAny(ref.getObjectId());
```

Comment:
Yep, refs can point to different things but `parseCommit()` throws `IncorrectObjectTypeException` if it's something other than a commit from what I can see in the docs. Doesn't matter that much, I just thought it would be more readable that way.

Reply:
cool, I will change it